File: frontend/src/services/constraints/nodeDataBuilder/foreignKey.ts
/**
 * @fileoverview ForeignKey builder — FK 特殊双引用处理
 *
 * FK 约束有两个列引用（sourceRef + targetRef），
 * 需要创建两条边：一条约束输入边 + 一条 FK 展示边。
 */

import type { BuildInput, BuildResult } from './types'
import { registerBuilder } from './registry'

function buildForeignKey(input: BuildInput): BuildResult {
  const { fkRefs, tableName, configName, nodeId, mode, saveState, embedded } = input

  if (!fkRefs) {
    // 缺少 FK 引用，降级为空数据
    return {
      nodeData: {
        configName,
        saveState: saveState || (mode === 'connect' ? 'draft' : 'saved'),
      },
      edgeDescriptors: [],
    }
  }

  const nodeData: Record<string, unknown> = {
    configName,
    sourceTable: fkRefs.source.columnName ? tableName : '',
    sourceColumn: fkRefs.source.columnName,
    targetTable: fkRefs.target.columnName
      ? input.refs?.to_table_name
        ? String(input.refs.to_table_name)
        : ''
      : '',
    targetColumn: fkRefs.target.columnName,
    validationStatus: 'idle',
    validationErrors: [],
    sourceRef: { nodeId: fkRefs.source.nodeId, columnId: fkRefs.source.columnId },
    targetRef: { nodeId: fkRefs.target.nodeId, columnId: fkRefs.target.columnId },
    config: {
      ruleType: 'EXIST_IN',
      targetNodeId: fkRefs.target.nodeId,
      targetColumn: fkRefs.target.columnName,
    },
    saveState: saveState || (mode === 'connect' ? 'draft' : 'saved'),
  }

  if (embedded) {
    nodeData.embedded = true
  }

  const edgeDescriptors = []

  // 约束输入边：Schema → FK 节点
  if (fkRefs.source.nodeId && fkRefs.source.columnId) {
    edgeDescriptors.push({
      kind: 'constraint' as const,
      sourceNodeId: fkRefs.source.nodeId,
      targetNodeId: nodeId,
      columnId: fkRefs.source.columnId,
    })
  }

  // FK 展示边：FK 节点 → 目标 Schema
  if (fkRefs.target.nodeId && fkRefs.target.columnId) {
    const label = [fkRefs.source.columnName, fkRefs.target.columnName].filter(Boolean).length
      ? `${fkRefs.source.columnName} → ${fkRefs.target.columnName}`
      : 'ForeignKey'

    edgeDescriptors.push({
      kind: 'fkDisplay' as const,
      sourceNodeId: nodeId,
      targetNodeId: fkRefs.target.nodeId,
      columnId: fkRefs.target.columnId,
      sourceHandle: `source-output-${nodeId}`,
      targetHandle: `source-right-${fkRefs.target.columnId}`,
      extra: {
        label,
        edgeId: `fk-${fkRefs.source.nodeId}-${fkRefs.target.nodeId}-${nodeId}`,
        fromTableId: fkRefs.source.nodeId,
        toTableId: fkRefs.target.nodeId,
        fromColumnId: fkRefs.source.columnId,
        toColumnId: fkRefs.target.columnId,
        constraintId: nodeId,
      },
    })
  }

  return { nodeData, edgeDescriptors }
}

registerBuilder('foreignKey', buildForeignKey)

File: backend/app/shared/core/project/constraint/builders/foreign_key.py
"""@fileoverview ForeignKey 约束构建器

双向引用（from/to 两端表+列），与其他单端约束不同。
"""

from __future__ import annotations

from .base import BuilderInput, BuilderResult
from .registry import register_builder


@register_builder("ForeignKey")
def build_foreign_key(inp: BuilderInput) -> BuilderResult:
    """ForeignKey: refs {from_table_id, from_column_id, to_table_id, to_column_id}。"""
    refs = inp.refs
    from_table_id = refs.get("from_table_id")
    from_col_id = refs.get("from_column_id")
    to_table_id = refs.get("to_table_id")
    to_col_id = refs.get("to_column_id")

    if from_table_id is None or to_table_id is None:
        return {}, "缺少必要的表引用"

    from_col_name = inp.column_name_by_table_id.get(from_table_id, {}).get(str(from_col_id))
    to_col_name = inp.column_name_by_table_id.get(to_table_id, {}).get(str(to_col_id))

    if from_col_name is None:
        return {}, f"引用的列 '{from_col_id}' 不存在于表 '{from_table_id}' 中"
    if to_col_name is None:
        return {}, f"引用的列 '{to_col_id}' 不存在于表 '{to_table_id}' 中"

    return {
        "from_table": from_table_id,
        "from_column": from_col_name,
        "to_table": to_table_id,
        "to_column": to_col_name,
    }, None
